flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryDefaultWriter.java

```diff
@@ -40,20 +47,67 @@
  *
  * @param <IN> Type of records to be written to BigQuery.
  */
-public class BigQueryDefaultWriter<IN> implements SinkWriter<IN> {
+public class BigQueryDefaultWriter<IN> extends BaseWriter<IN> {
 
+    public BigQueryDefaultWriter(
+            int subtaskId,
+            BigQueryConnectOptions connectOptions,
+            BigQuerySchemaProvider schemaProvider,
+            BigQueryProtoSerializer serializer,
+            String tablePath) {
+        super(subtaskId, connectOptions, schemaProvider, serializer);
+        streamName = String.format("%s/streams/_default", tablePath);
+    }
+
+    /** Accept record for writing to BigQuery table. */
     @Override
-    public void write(IN element, Context context) throws IOException, InterruptedException {
-        throw new UnsupportedOperationException("write method is not supported");
+    public void write(IN element, Context context) {
+        try {
+            ByteString protoRow = getProtoRow(element);
+            if (!fitsInAppendRequest(protoRow)) {
+                validateAppendResponses(false);
+                append();
+            }
+            addToAppendRequest(protoRow);
+        } catch (BigQuerySerializationException e) {
+            logger.error(String.format("Unable to serialize record %s. Dropping it!", element), e);
+        }
     }
 
+    /** Asynchronously append to BigQuery table's default stream. */
     @Override
-    public void flush(boolean endOfInput) throws IOException, InterruptedException {
-        throw new UnsupportedOperationException("flush method is not supported");
+    void sendAppendRequest() {
+        if (streamWriter == null) {
+            streamWriter = createStreamWriter(true);
+        }
+        ApiFuture responseFuture = streamWriter.append(protoRowsBuilder.build());
+        appendResponseFuturesQueue.add(responseFuture);
     }
 
+    /** Throws a RuntimeException if an error is found with append response. */
     @Override
-    public void close() {
-        throw new UnsupportedOperationException("close method is not supported");
+    void validateAppendResponse(ApiFuture<AppendRowsResponse> appendResponseFuture) {
```

Comment:
Do you anticipate any difference in this method for the exactly-once writer? If not, then this could go on the base class?

Reply:
Yes there will be some differences for the exactly-once writer, such as handling of offset related errors.